File: src/logic/agents/AgentMovePhys.ts
import { Agent, AgentState, STUCK_DANGER_2 } from "./Agent";
import { GameState } from "../GameState";
import { add, dot, length, length_sq, add_, set, scale_, set_, distance_sq, subtract_, normalize_, lerp, cvt } from "../core/math";
import { raycastPoint } from "../Raycasting";
import { findTriangle, isPointInNavmesh } from "../navmesh/NavUtils";
import { NavConst } from "./NavConst";


let desiredVelocity = { x: 0, y: 0 };
let moveVector = { x: 0, y: 0 };
let velocityDiff = { x: 0, y: 0 };
let directionToCorner = { x: 0, y: 0 };
let endPoint = { x: 0, y: 0 };
let normVelocity = { x: 0, y: 0 };
let endPointForRecast = { x: 0, y: 0 };
let wallVector = { x: 0, y: 0 };
let wallNormal = { x: 0, y: 0 };
let impulse = { x: 0, y: 0 };
let tempScaled = { x: 0, y: 0 };
let finalAccelDirection = { x: 0, y: 0 };
let corner1ToCorner2 = { x: 0, y: 0 };
let normCorner1ToCorner2 = { x: 0, y: 0 };
const MAX_SPEED = 999999;

export function updateAgentPhys(agent: Agent, deltaTime: number, gs: GameState): void {
    const navmesh = gs.navmesh;
    set_(agent.lastCoordinate, agent.coordinate);

    if (length_sq(agent.velocity) < 0.001) {
        agent.velocity.x = 0;
        agent.velocity.y = 0;
    }

    const resistanceFactor = 1 - agent.resistance;
    const frameRateAdjustedResistance = Math.pow(resistanceFactor, deltaTime);

    // Desired velocity
    let desiredMagnitude = 1;
    set_(directionToCorner, agent.nextCorner);
    subtract_(directionToCorner, agent.coordinate);
    let dstToCorner = length(directionToCorner);
    scale_(directionToCorner, dstToCorner > 0.01 ? 1 / dstToCorner : 0);
    set_(desiredVelocity, directionToCorner);
    if (agent.state === AgentState.Traveling || agent.state === AgentState.Escaping) {
        let slowDownStrength = 1 / 8 / agent.resistance / agent.resistance;
        slowDownStrength *= lerp(0.5, 2, agent.intelligence);
        let slowBeforeCornerDst = agent.maxSpeed * 0.25;
        let slowBeforeCornerSpeed = agent.maxSpeed;
        
        if (dstToCorner < slowBeforeCornerDst && agent.numValidCorners >= 2) {
            set_(corner1ToCorner2, agent.nextCorner2);
            subtract_(corner1ToCorner2, agent.nextCorner);
            normalize_(corner1ToCorner2);                
            set_(normVelocity, agent.velocity);
            normalize_(normVelocity);
            let turnAlignment = dot(normVelocity, corner1ToCorner2);
            turnAlignment = (turnAlignment + 1) * 0.5
            turnAlignment = turnAlignment * turnAlignment * turnAlignment;
            slowBeforeCornerDst *= lerp(1, 0, turnAlignment);
            slowBeforeCornerSpeed *= lerp(slowDownStrength, 1, turnAlignment);
        }
        
        if (dstToCorner > slowBeforeCornerDst)
            desiredMagnitude = agent.maxSpeed;
        else {
            const minSpeed = agent.numValidCorners == 1 ? agent.arrivalDesiredSpeed * agent.maxSpeed : slowBeforeCornerSpeed
            desiredMagnitude = lerp(minSpeed, agent.maxSpeed, dstToCorner / slowBeforeCornerDst);
            
        }
    }
    else {
        desiredMagnitude = 0;
    }
    desiredMagnitude /= frameRateAdjustedResistance;
    const stuckFactor = agent.stuckRating / STUCK_DANGER_2;
    desiredMagnitude *= cvt(stuckFactor * stuckFactor, 0, 1, 1, 0.5);
    scale_(desiredVelocity, desiredMagnitude);

    set_(agent.debug_desiredVelocity, desiredVelocity);
    
    // Dumb agents: accelerate toward desired velocity (A)
    // Smart agents: accelerate toward velocity difference (A - B)
    set_(velocityDiff, desiredVelocity);
    subtract_(velocityDiff, agent.velocity);
    set_(agent.debug_velocityDiff, velocityDiff);
    let effectiveInt = length_sq(desiredVelocity) > 0.1 ? agent.intelligence : 1;
    
    // Blend: (1-intelligence) * A + intelligence * (A-B)
    set_(finalAccelDirection, directionToCorner);
    let requiredAddition = desiredMagnitude - dot(agent.velocity, directionToCorner);
    scale_(finalAccelDirection, requiredAddition * (1 - effectiveInt));
    scale_(velocityDiff, effectiveInt);
    add_(finalAccelDirection, velocityDiff);
    
    const diffLn = length(finalAccelDirection);
    const accelThisFrame = Math.min(diffLn, agent.accel * deltaTime);
    scale_(finalAccelDirection, accelThisFrame / (diffLn > 0.001 ? diffLn : 1));
    
    if (deltaTime > 0.0001) {
        set_(agent.lastAppliedAccel, finalAccelDirection);
        scale_(agent.lastAppliedAccel, 1 / deltaTime);
    } else {
        set(agent.lastAppliedAccel, 0, 0);
    }

    add_(agent.velocity, finalAccelDirection);

    // Apply resistance
    agent.velocity.x *= frameRateAdjustedResistance;
    agent.velocity.y *= frameRateAdjustedResistance;

    set_(moveVector, agent.velocity);
    scale_(moveVector, deltaTime);
    const moveLnSq = length_sq(moveVector);

    // --- Escaping Movement: Beeline to last valid spot, no collision. ---
    if (agent.state === AgentState.Escaping) {
        const distanceToTargetSq = distance_sq(agent.nextCorner, agent.coordinate);
        
        // If the next move would overshoot the target, just snap to it.
        if (moveLnSq >= distanceToTargetSq) {
            set_(agent.coordinate, agent.lastValidPosition);
            set(agent.velocity, 0, 0);
            // agent.currentTri = agent.lastValidTri;
        } else {
            add_(agent.coordinate, moveVector);
        }
    } 
    // --- Standard Movement with Collision ---
    else {
        if (deltaTime > 0 && moveLnSq > 0.0001) {
            // Use temporary variables to avoid allocations
            set_(endPoint, agent.coordinate);
            add_(endPoint, moveVector);
            
            set_(normVelocity, agent.velocity);
            normalize_(normVelocity);
            
            set_(tempScaled, normVelocity);
            scale_(tempScaled, 0.45);
            set_(endPointForRecast, endPoint);
            add_(endPointForRecast, tempScaled);
            
            const raycastResult = raycastPoint(navmesh, agent.coordinate, endPointForRecast, agent.currentTri, undefined);

            if (raycastResult.hitP1 && raycastResult.hitP2) {
                if (!agent.wallContact) {
                    agent.wallContact = true;
                }
                agent.stuckRating += NavConst.STUCK_HIT_WALL;
                set_(wallVector, raycastResult.hitP2);
                subtract_(wallVector, raycastResult.hitP1);
                
                set(wallNormal, -wallVector.y, wallVector.x);
                normalize_(wallNormal);
                
                if (dot(wallNormal, normVelocity) > 0) {
                    scale_(wallNormal, -1);
                }

                const normalVelocityComponent = dot(agent.velocity, wallNormal);
                agent.velocity.x -= normalVelocityComponent * wallNormal.x * 1.45;
                agent.velocity.y -= normalVelocityComponent * wallNormal.y * 1.45;
                set_(moveVector, agent.velocity);
                scale_(moveVector, deltaTime);                
                add_(agent.coordinate, moveVector);
            } else {
                if (agent.wallContact) {
                    agent.wallContact = false;
                }
                set_(agent.coordinate, endPoint);
            }
        }
    }
    
    const oldTri = agent.currentTri;
    const newTriRaw = findTriangle(agent.coordinate, gs.navmesh, agent.currentTri);
    if (oldTri !== newTriRaw && newTriRaw !== -1) {
        if (newTriRaw >= gs.navmesh.walkable_triangle_count) {
            console.log(`TS agent ${agent.id} changed to unwalkable triangle from ${oldTri} to ${newTriRaw}. Walkable limit: ${gs.navmesh.walkable_triangle_count}`);
        }
    }
    const newTri = isPointInNavmesh(agent.coordinate, gs.navmesh, agent.currentTri);
    if (newTri !== -1) {
        agent.currentTri = newTri;
        set_(agent.lastValidPosition, agent.coordinate);
        agent.lastValidTri = newTri;
    } else {
        agent.currentTri = -1;
        
    }

    // // Occasional random impulse for testing path correction
    // if (agent.state !== AgentState.Escaping && Math.random() < 0.05) {
    //     set(impulse, (Math.random() - 0.5) * 130, (Math.random() - 0.5) * 130);
    //     add_(agent.velocity, impulse);
    // }
} 

File: src/wasm/agent_move_phys.cpp
#include "agent_move_phys.h"
#include "math_utils.h"
#include "raycasting.h"
#include "nav_utils.h"
#include "data_structures.h" // brings in constants_layout.h macros
#include <cmath>
#include "constants_layout.h"
#include <cstdio>

extern Navmesh g_navmesh;
extern float g_sim_time;
extern std::vector<uint8_t> g_wall_contact;

void update_agent_phys(int idx, float deltaTime) {
  agent_data.last_coordinates[idx] = agent_data.positions[idx];

  if (math::length_sq(agent_data.velocities[idx]) < 0.001f) {
    agent_data.velocities[idx] = {0.0f, 0.0f};
  }

  const float resistance = agent_data.resistances[idx];
  const float frameRateAdjustedResistance = pow(1.0f - resistance, deltaTime);

  // Desired velocity calculation
  Point2 directionToCorner = agent_data.next_corners[idx] - agent_data.positions[idx];
  float dstToCorner = math::length(directionToCorner);
  if (dstToCorner > 0.01f) {
    directionToCorner /= dstToCorner;
  } else {
    directionToCorner = {0,0};
  }
  
  Point2 desiredVelocity = directionToCorner;
  float desiredMagnitude = 0;

  if (agent_data.states[idx] == AgentState::Traveling || agent_data.states[idx] == AgentState::Escaping) {
    float maxSpeed = agent_data.max_speeds[idx];
    float intelligence = agent_data.intelligences[idx];
    
    float slowDownStrength = 1.0f / 8.0f / resistance / resistance;
    slowDownStrength *= math::lerp(0.5f, 2.0f, intelligence);
    float slowBeforeCornerDst = maxSpeed * 0.25f;
    float slowBeforeCornerSpeed = maxSpeed;

    if (dstToCorner < slowBeforeCornerDst && agent_data.num_valid_corners[idx] >= 2) {
      Point2 corner1ToCorner2 = agent_data.next_corners2[idx] - agent_data.next_corners[idx];
      math::normalize_inplace(corner1ToCorner2);
      
      Point2 normVelocity = agent_data.velocities[idx];
      math::normalize_inplace(normVelocity);
      
      float turnAlignment = math::dot(normVelocity, corner1ToCorner2);
      turnAlignment = (turnAlignment + 1.0f) * 0.5f;
      turnAlignment = turnAlignment * turnAlignment * turnAlignment;
      slowBeforeCornerDst *= math::lerp(1.0f, 0.0f, turnAlignment);
      slowBeforeCornerSpeed *= math::lerp(slowDownStrength, 1.0f, turnAlignment);
    }

    if (dstToCorner > slowBeforeCornerDst) {
      desiredMagnitude = maxSpeed;
    } else {
      const float minSpeed = agent_data.num_valid_corners[idx] == 1 
        ? agent_data.arrival_desired_speeds[idx] * maxSpeed 
        : slowBeforeCornerSpeed;
      desiredMagnitude = math::lerp(minSpeed, maxSpeed, dstToCorner / slowBeforeCornerDst);
    }
  } else {
    desiredMagnitude = 0;
  }

  desiredMagnitude /= frameRateAdjustedResistance;
  const float stuckFactor = agent_data.stuck_ratings[idx] / STUCK_DANGER_2;
  desiredMagnitude *= math::cvt(stuckFactor * stuckFactor, 0.0f, 1.0f, 1.0f, 0.5f);
  
  desiredVelocity = directionToCorner * desiredMagnitude;

  Point2 velocityDiff = desiredVelocity - agent_data.velocities[idx];
  
  float effectiveInt = math::length_sq(desiredVelocity) > 0.1f ? agent_data.intelligences[idx] : 1.0f;

  Point2 finalAccelDirection = directionToCorner;
  float requiredAddition = desiredMagnitude - math::dot(agent_data.velocities[idx], directionToCorner);
  finalAccelDirection *= requiredAddition * (1.0f - effectiveInt);
  velocityDiff *= effectiveInt;
  finalAccelDirection += velocityDiff;

  const float diffLn = math::length(finalAccelDirection);
  const float accelThisFrame = fmin(diffLn, agent_data.accels[idx] * deltaTime);
  if (diffLn > 0.001f) {
    finalAccelDirection *= (accelThisFrame / diffLn);
  } else {
    finalAccelDirection *= 0;
  }

  agent_data.velocities[idx] += finalAccelDirection;

  agent_data.velocities[idx] *= frameRateAdjustedResistance;

  Point2 moveVector = agent_data.velocities[idx] * deltaTime;
  const float moveLnSq = math::length_sq(moveVector);

  if (agent_data.states[idx] == AgentState::Escaping) {
    const float distanceToTargetSq = math::distance_sq(agent_data.next_corners[idx], agent_data.positions[idx]);
    if (moveLnSq >= distanceToTargetSq) {
      agent_data.positions[idx] = agent_data.last_valid_positions[idx];
      agent_data.velocities[idx] = {0.0f, 0.0f};
    } else {
      agent_data.positions[idx] += moveVector;
    }
  } else {
    if (deltaTime > 0.0f && moveLnSq > 0.0001f) {
      Point2 endPoint = agent_data.positions[idx] + moveVector;
      Point2 normVelocity = math::normalize(agent_data.velocities[idx]);
      Point2 endPointForRecast = endPoint + (normVelocity * 0.45f);
      
      auto raycastResult = raycastPoint(agent_data.positions[idx], endPointForRecast, agent_data.current_tris[idx]);

      if (std::get<2>(raycastResult)) {
        if (!g_wall_contact.empty() && g_wall_contact[idx] == 0) {
          g_wall_contact[idx] = 1;
        }
        agent_data.stuck_ratings[idx] += STUCK_HIT_WALL;
        Point2 wallVector = std::get<1>(raycastResult) - std::get<0>(raycastResult);
        Point2 wallNormal = {-wallVector.y, wallVector.x};
        math::normalize_inplace(wallNormal);

        if (math::dot(wallNormal, normVelocity) > 0) {
          wallNormal *= -1.0f;
        }

        const float normalVelocityComponent = math::dot(agent_data.velocities[idx], wallNormal);
        agent_data.velocities[idx].x -= normalVelocityComponent * wallNormal.x * 1.45;
        agent_data.velocities[idx].y -= normalVelocityComponent * wallNormal.y * 1.45;
        moveVector = agent_data.velocities[idx] * deltaTime;
        agent_data.positions[idx] += moveVector;
      } else {
        if (!g_wall_contact.empty() && g_wall_contact[idx] == 1) {
          g_wall_contact[idx] = 0;
        }
        agent_data.positions[idx] = endPoint;
      }
    }
  }
  
  int oldTri = agent_data.current_tris[idx];
  int newTri = is_point_in_navmesh(agent_data.positions[idx], agent_data.current_tris[idx]);
  if (oldTri != newTri && newTri != -1) {
    if (newTri >= g_navmesh.walkable_triangle_count) {
      printf("WA agent %d changed to unwalkable triangle from %d to %d. Walkable limit: %d\n", idx, oldTri, newTri, g_navmesh.walkable_triangle_count);
    }
  }
  if (newTri != -1) {
    agent_data.current_tris[idx] = newTri;
    agent_data.last_valid_positions[idx] = agent_data.positions[idx];
    agent_data.last_valid_tris[idx] = newTri;
  } else {
    agent_data.current_tris[idx] = -1;
  }
}
